# 1. load_digits : 손글씨 분류

In [10]:
#모듈 import
from sklearn.datasets import load_digits 
from sklearn.model_selection import train_test_split  #데이터셋 분리 모듈
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

#데이터 준비
digits = load_digits()
digits_data = digits.data 
digits_label = digits.target  #digits_label변수에 digits 데이터의 target저장

In [11]:
#Target Names출력
digits.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [12]:
#데이터 Describe
print(digits.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

In [13]:
#trian, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(digits_data,  #feature
                                                    digits_label,  #label
                                                    test_size=0.2,   #test dataset 크기
                                                    random_state=7)  #데이터 분리 시 랜덤성

In [14]:
#decision tree
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.81      0.81      0.81        42
           2       0.79      0.82      0.80        40
           3       0.79      0.91      0.85        34
           4       0.83      0.95      0.89        37
           5       0.90      0.96      0.93        28
           6       0.84      0.93      0.88        28
           7       0.96      0.82      0.89        33
           8       0.88      0.65      0.75        43
           9       0.78      0.78      0.78        32

    accuracy                           0.86       360
   macro avg       0.86      0.86      0.86       360
weighted avg       0.86      0.86      0.85       360

0.8555555555555555


In [15]:
#random forest
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=32) 
random_forest.fit(X_train, y_train) 
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.93      1.00      0.97        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       0.93      1.00      0.96        37
           5       0.90      0.96      0.93        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       1.00      0.84      0.91        43
           9       0.94      0.94      0.94        32

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.97      0.96      0.96       360

0.9638888888888889


In [16]:
#svm
from sklearn import svm

svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      1.00      0.98        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        37
           5       0.93      1.00      0.97        28
           6       1.00      1.00      1.00        28
           7       1.00      1.00      1.00        33
           8       1.00      0.93      0.96        43
           9       1.00      0.97      0.98        32

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360

0.9888888888888889


In [17]:
#sgd classifier
from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.88      0.88      0.88        42
           2       1.00      0.95      0.97        40
           3       0.82      0.91      0.86        34
           4       1.00      0.97      0.99        37
           5       0.90      1.00      0.95        28
           6       0.96      0.93      0.95        28
           7       0.97      0.97      0.97        33
           8       0.89      0.91      0.90        43
           9       0.97      0.88      0.92        32

    accuracy                           0.94       360
   macro avg       0.94      0.94      0.94       360
weighted avg       0.94      0.94      0.94       360

0.9361111111111111


In [28]:
#Logistic Regression
from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      0.95      0.95        42
           2       0.98      1.00      0.99        40
           3       0.94      0.97      0.96        34
           4       0.97      1.00      0.99        37
           5       0.82      0.96      0.89        28
           6       1.00      0.96      0.98        28
           7       0.97      0.97      0.97        33
           8       0.92      0.81      0.86        43
           9       0.97      0.91      0.94        32

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360

0.9527777777777777


/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


# 모델 평가
decision tree:85.5%, random forest:96.3%, svm:98.8%, sgd classifier:93.6%, Logistic Regression:95.2%

svm 모델 사용시 accuracy가 98.8%로 가장 높게 도출되었다.


# 2. load_wine : 와인 분류

In [29]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

#데이터 준비
wine = load_wine()
wine_data = wine.data 
wine_label = wine.target

In [30]:
#Target Names 출력
wine.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [31]:
#데이터 Describe
print(digits.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

In [32]:
#trian, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(wine_data,  #feature
                                                    wine_label,  #label
                                                    test_size=0.2,   #test dataset 크기
                                                    random_state=7)  #데이터 분리 시 랜덤성

In [33]:
#decision tree
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       0.89      1.00      0.94        17
           2       1.00      0.83      0.91        12

    accuracy                           0.94        36
   macro avg       0.96      0.94      0.95        36
weighted avg       0.95      0.94      0.94        36

0.9444444444444444


In [34]:
#random forest
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=32) 
random_forest.fit(X_train, y_train) 
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       1.00      1.00      1.00        17
           2       1.00      1.00      1.00        12

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36

1.0


In [35]:
#svm
from sklearn import svm

svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       0.86      0.86      0.86         7
           1       0.58      0.88      0.70        17
           2       0.33      0.08      0.13        12

    accuracy                           0.61        36
   macro avg       0.59      0.61      0.56        36
weighted avg       0.55      0.61      0.54        36

0.6111111111111112


In [36]:
#sgd classifier
from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      0.57      0.73         7
           1       0.53      1.00      0.69        17
           2       0.00      0.00      0.00        12

    accuracy                           0.58        36
   macro avg       0.51      0.52      0.47        36
weighted avg       0.45      0.58      0.47        36

0.5833333333333334


/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [37]:
#Logistic Regression
from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      0.86      0.92         7
           1       0.94      1.00      0.97        17
           2       1.00      1.00      1.00        12

    accuracy                           0.97        36
   macro avg       0.98      0.95      0.96        36
weighted avg       0.97      0.97      0.97        36

0.9722222222222222


/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


# 모델 평가
decision tree:94.4%, random forest:100%, svm:61.1%, sgd classifier:58.3% Logistic Regression:97.2%

 random forest모델 사용시 accurac가 100%로 가장 높게 도출되었다.
 # 100% 의심스럽긴 하지만 계속 높게나왔다

# 3. load_breast_cancer : 유방암 여부 진단

In [38]:
#모듈 import
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

#데이터 준비
breast_cancer = load_breast_cancer()
breast_cancer_data = breast_cancer.data 
breast_cancer_label = breast_cancer.target

In [39]:
#Target Names 출력
breast_cancer.target_names

array(['malignant', 'benign'], dtype='<U9')

In [40]:
#데이터 Describe
print(breast_cancer.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

In [41]:
#trian, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data,  #feature
                                                    breast_cancer_label,  #label
                                                    test_size=0.2,   #test dataset 크기
                                                    random_state=7)  #데이터 분리 시 랜덤성

In [42]:
#decision tree
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       0.92      0.82      0.87        40
           1       0.91      0.96      0.93        74

    accuracy                           0.91       114
   macro avg       0.91      0.89      0.90       114
weighted avg       0.91      0.91      0.91       114

0.9122807017543859


In [43]:
#random forest
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=32) 
random_forest.fit(X_train, y_train) 
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        40
           1       1.00      1.00      1.00        74

    accuracy                           1.00       114
   macro avg       1.00      1.00      1.00       114
weighted avg       1.00      1.00      1.00       114

1.0


In [44]:
#svm
from sklearn import svm

svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      0.72      0.84        40
           1       0.87      1.00      0.93        74

    accuracy                           0.90       114
   macro avg       0.94      0.86      0.89       114
weighted avg       0.92      0.90      0.90       114

0.9035087719298246


In [45]:
#sgd classifier
from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      0.70      0.82        40
           1       0.86      1.00      0.92        74

    accuracy                           0.89       114
   macro avg       0.93      0.85      0.87       114
weighted avg       0.91      0.89      0.89       114

0.8947368421052632


In [46]:
#Logistic Regression
from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      0.85      0.92        40
           1       0.93      1.00      0.96        74

    accuracy                           0.95       114
   macro avg       0.96      0.93      0.94       114
weighted avg       0.95      0.95      0.95       114

0.9473684210526315


/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


# 모델 평가
decision tree:91.2%, random forest:100%, svm:90.3%, sgd classifier:89.4% Logistic Regression:94.7%

random forest 모델 사용시 accuracy가 100%로 가장 높게 도출되었다.

# E-02 회고

• 이번 프로젝트는 두 번째라 가위바위보 보다는 어떻게 해야할지 미미하지만 어느정도 감이라는게 잡혀서 조금 원활하게 진행할 수 있었던 것 같습니다.
하지만 accuracy에서 100%가 나왔을 때는 왜 100%가 나오는지 몰라서 몇 번이고 돌려보고 구글링을 해봤지만 아직도 잘 모르겠습니다.

• 프로젝트를 진행하면서 모호한 점 당장 위에 보이는 /opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

이 부분이 왜 나오는지 아직까지 모르겠습니다. 구글링을 통해 알아보니 AlphaFold를 설치하려고 하는데 빌드하는 동안 conda오류가 발생하는 거라는데 아무리봐도 아닌 것 같습니다.

• 루브릭 평가 지표를 맞추기 위해 시도한 것들 평가테스트가 정상적으로 나올 수 있도록 중간에 import가 안 된 부분을 찾아서 추가했습니다. 하지만ㅎㅎㅎ 100%라는 결과가 정상적으로 도출된 것이 맞는지 아직 의심이 듭니다.

• 자기 다짐
저번 E-01를 진행할 때와 다르게 조금ㅎㅎㅎ 아주 조금 방향을 잡아서 하는 방법을 알게되었고 구글링을 통해 얻은 자료를 나름 과감히 사용하게 되었으나 아직도 헐레벌떡 하는 느낌이 강합니다... 조금씩 증진해가며 능숙하게 프로젝트를 진행할 수 있도록 만들어야겠습니다:)